runtime/freelist.c

```diff
@@ -534,60 +348,1404 @@ header_t *caml_fl_merge_block (value bp)
+    while (1){
+      if (p == Val_NULL){
+        unset_map (i);
+        break;
+      }
+      if (Color_val (p) == Caml_blue) break;
+      CAMLassert (Color_val (p) == Caml_white);
+      caml_fl_cur_wsz -= Whsize_val (p);
+      p = Next_small (p);
+    }
+    bf_small_fl[i].free = p;
+    /* Set the merge pointer to its initial value */
+    bf_small_fl[i].merge = &bf_small_fl[i].free;
+  }
+}
+
+static void bf_reset (void)
+{
+  mlsize_t i;
+
+  for (i = 1; i <= BF_NUM_SMALL; i++){
+    bf_small_fl[i].free = Val_NULL;
+    bf_small_fl[i].merge = &(bf_small_fl[i].free);
+  }
+  bf_small_map = 0;
+  bf_large_tree = NULL;
+  bf_large_least = NULL;
+  caml_fl_cur_wsz = 0;
+  bf_init_merge ();
+}
+
+static header_t *bf_merge_block (value bp, char *limit)
+{
+  value start;
+  value cur;
+  mlsize_t wosz;
+
+  CAMLassert (Color_val (bp) == Caml_white);
+  /* Find the starting point of the current run of free blocks. */
+  if (caml_fl_merge != Val_NULL && Next_in_mem (caml_fl_merge) == bp){
+    start = caml_fl_merge;
+    CAMLassert (Color_val (start) == Caml_blue);
+    bf_remove (start);
+  }else{
+    start = bp;
+  }
+  cur = bp;
+  while (1){
+    /* This slightly convoluted loop is just going over the run of
+       white or blue blocks, doing the right thing for each color, and
+       stopping on a gray or black block or when limit is passed.
+       It is convoluted because we start knowing that the first block
+       is white. */
+  white:
+    if (Tag_val (cur) == Custom_tag){
+      void (*final_fun)(value) = Custom_ops_val(cur)->finalize;
+      if (final_fun != NULL) final_fun(cur);
+    }
+    caml_fl_cur_wsz += Whsize_val (cur);
+  next:
+    cur = Next_in_mem (cur);
+    if ((char *) cur > limit) goto end_of_run;
+    switch (Color_val (cur)){
+    case Caml_white: goto white;
+    case Caml_blue: bf_remove (cur); goto next;
+    case Caml_gray:
+    case Caml_black:
+      goto end_of_run;
+    }
+  }
+ end_of_run:
+  wosz = Wosize_whsize ((value *) cur - (value *) start);
+#ifdef DEBUG
+  {
+    value *p;
+    for (p = (value *) start; p < (value *) Hp_val (cur); p++){
+      *p = Debug_free_major;
+    }
+  }
+#endif
+  while (wosz > Max_wosize){
+    Hd_val (start) = Make_header (Max_wosize, 0, Caml_blue);
+    bf_insert_sweep (start);
+    start = Next_small (start);
+    wosz -= Whsize_wosize (Max_wosize);
+  }
+  if (wosz > 0){
+    Hd_val (start) = Make_header (wosz, 0, Caml_blue);
+    bf_insert_sweep (start);
+  }else{
+    Hd_val (start) = Make_header (0, 0, Caml_white);
+    caml_fl_cur_wsz -= Whsize_wosize (0);
+  }
+  FREELIST_DEBUG_bf_check ();
+  return Hp_val (cur);
+}
+
+static void bf_add_blocks (value bp)
+{
+  while (bp != Val_NULL){
+    value next = Next_small (bp);
+    mlsize_t wosz = Wosize_val (bp);
+
+    if (wosz > BF_NUM_SMALL){
+      caml_fl_cur_wsz += Whsize_wosize (wosz);
+      bf_insert_block ((large_free_block *) bp);
+    }else{
+      Hd_val (bp) = Whitehd_hd (Hd_val (bp));
```

Comment:
Good catch. Note that it would be very hard to trigger this bug: you need a 32-bit *debug* runtime, a heap extension by a size 2 to 32 modulo Max_whsize, a stray pointer inside the newly-allocated block, and a heap check to trigger the assertion.
Fixed.